nx/lib/nx/shared.ex

```diff
@@ -387,4 +393,27 @@ defmodule Nx.Shared do
             "#{inspect(struct1)} and #{inspect(struct2)}. You may need to call Nx.backend_transfer/1 " <>
             "(or Nx.backend_copy/1) on one or both of them to transfer them to a common implementation"
   end
+
+  @doc """
+  Used to define an Nx callback with an optional implementation.
+
+  The given body is used as the default implementation otherwise.
+  """
+  def default_implementation(function_name, args, default_impl)
+      when is_atom(function_name) and is_list(args) and is_function(default_impl) do
+    arity = length(args)
+
+    backend = impl!(args)
+
+    cond do
+      backend == Nx.Defn.Expr ->
+        apply(default_impl, args)
```

Comment:
@josevalim I'm a little bit confused as to why, but passing the Expr tensors to the default implementation got the broken tests to pass (some of which were grad tests as well)

Reply:
I don't know what to say except that this implementation looks correct to me. :D However, it is only part 1, because if you use `defn`, we won't currently dispatch to the optional implementation. :) But I think this is mergeable as is.